### PR TITLE
adds docker coding guideline

### PR DESCRIPTION
**Pull request by venturel**
As per https://github.com/skyscrapers/engineering/issues/220 adding some docker best practice

### CODE REVIEW

File: coding_guidelines/docker.md [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+#### FROM
+
+Whenever possible, use current official images as the basis for your images. We recommend the Alpine image as it is tightly controlled and small in size (currently under 5 MB), while still being a full Linux distribution.
+
+#### RUN
+
+Split long or complex RUN statements on multiple lines separated with backslashes to make your Dockerfile more readable, understandable, and maintainable.
+
+#### APT-GET
```
Comment (iuriaranda) [inline coding_guidelines/docker.md]: I would add a section for `apk` as well, as we encourage to use Alpine. Something in the lines of the apt-get section:

- don't cache the repo index
- remove build packages after they're no longer needed

Good example: https://github.com/skyscrapers/docker-images/blob/master/aws-nuke/Dockerfile
curl is only installed to fetch some dependencies and then removed

File: coding_guidelines/docker.md [hunk 2]
```diff
@@ -3,7 +3,7 @@
 This document covers recommended best practices and methods for building efficient images.
 
 You can find more details in the (official Docker documentation)[https://docs.docker.com/develop/develop-images/dockerfile_best-practices/]
-Docker builds images automatically by reading the instructions from a Dockerfile -- a text file that contains all commands, in order, needed to build a given image. A Dockerfile adheres to a specific format and set of instructions which you can find at Dockerfile reference.
+Docker builds images automatically by reading the instructions from a Dockerfile -- a text file that contains all commands, in order, needed to build a given image. A Dockerfile adheres to a specific format and set of instructions which you can find at (Dockerfile reference)[https://docs.docker.com/engine/reference/builder/].
```
Comment (iuriaranda) [inline coding_guidelines/docker.md]: Links are constructed the other way around 😅 `[some text](https://heregoesthelink.com)`
And the same for the other links you added.

File: coding_guidelines/docker.md [hunk 3]
```diff
@@ -151,6 +153,11 @@ In addition, when you clean up the apt cache by removing /var/lib/apt/lists it r
 
 Official Debian and Ubuntu images automatically run apt-get clean, so explicit invocation is not required.
 
+#### APK-ADD
```
Comment (iuriaranda) [inline coding_guidelines/docker.md]: Can you clarify that `apk` is the package manager for Alpine. Might not be clear for everybody.

File: coding_guidelines/docker.md [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+### Decouple applications
+Each container should have only one concern. Decoupling applications into multiple containers makes it easier to scale horizontally and reuse containers. For instance, a web application stack might consist of three separate containers, each with its own unique image, to manage the web application, database, and an in-memory cache in a decoupled manner.
+
+Limiting each container to one process is a good rule of thumb, but it is not a hard and fast rule. For example, not only can containers be spawned with an init process, some programs might spawn additional processes of their own accord. For instance, Celery can spawn multiple worker processes, and Apache can create one process per request.
+
+
+### Minimize the number of layers
+In older versions of Docker, it was important that you minimized the number of layers in your images to ensure they were performant. The following features were added to reduce this limitation:
+
+Only the instructions RUN, COPY, ADD create layers. Other instructions create temporary intermediate images, and do not increase the size of the build.
```
[the remark this answers is not in the document]
Comment (venturel) [reply coding_guidelines/docker.md]: I am not sure what you are referring to

[the remark this answers is not in the document]
Comment (iuriaranda) [reply coding_guidelines/docker.md]: At the beginning of the document (line 7) you say this:

> Each instruction creates one layer:
> 
> FROM creates a layer from the ubuntu:18.04 Docker image.
> COPY adds files from your Docker client’s current directory.
> RUN builds your application with make.
> CMD specifies what command to run within the container.

But in here you say:

> Only the instructions RUN, COPY, ADD create layers

These two things are contradictory, so one of them is wrong I guess.

[the remark this answers is not in the document]
Comment (venturel) [reply coding_guidelines/docker.md]: I don't understand where the contradiction is. Each instruction of the example above what you pasted creates a layer. What's wrong with that?

**Comment by reyntjensw** [comment]
@venturel maybe a general question, wouldn't it be better to link to the sites the info comes from?

**Comment by venturel** [comment]
@reyntjensw it's in the document at line 5

**Comment by reyntjensw** [comment]
> @reyntjensw it's in the document at line 5

I know but I wouldn't copy / paste it in here because the documentation can change over time

File: coding_guidelines/docker.md [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+CMD python /app/app.py
+```
+
+Each instruction creates one layer:
+
+FROM creates a layer from the ubuntu:18.04 Docker image.
+COPY adds files from your Docker client’s current directory.
+RUN builds your application with make.
+CMD specifies what command to run within the container.
+When you run an image and generate a container, you add a new writable layer (the “container layer”) on top of the underlying layers. All changes made to the running container, such as writing new files, modifying existing files, and deleting files, are written to this thin writable container layer.
```
Comment (Komorebi-E) [inline coding_guidelines/docker.md]: When you run an image and generate a container from the Dockerfile, ...

**Pull request merged**